### PR TITLE
Fix 'Expr OP NDArr' by doing 'NDArr ROP Expr'

### PR DESCRIPTION
**Pull request by tias**
Fix 'Expr OP NDArr' by doing 'NDArr ROP Expr'

Fixes:
```
>>> x * y
NDVarArray([(a) * (y), (b) * (y), (c) * (y)], dtype=object)
>>> y * x
(y) * ([a b c])
```
from #1034

Curiously its actually a typing gap... that you can also get an
ndarray as second argument.

So with this now fixed, we might as well just go ahead and type all
these operators...?

### CODE REVIEW

**Comment by tias** [comment]
all of this is really pointing to that we should switch to ufuncs #341 ... but that seems a bit rushed to do now?

File: tests/test_expressions.py [hunk 1]
```diff
@@ -243,6 +243,38 @@ def test_mul_is_lhs_num(self):
         assert expr.is_lhs_num is False
         assert expr.args[0] is x and expr.args[1] is y
 
+    def test_scalar_expr_with_ndarray(self):
```
Comment (IgnaceBleukx) [inline tests/test_expressions.py:258]: Move these to `TestArrayExpression` class instead? Now they are in the tests for multiplication... 
Ideally also test for all of the other operators, not just mul.

**Review by IgnaceBleukx — APPROVED** [review]
Yeah absolutely we should have another look at ufuncs... But it will require lots of refactoring which is out of scope for the next release.

Current appraoch is a bit hacky of course, but fine as a patch for now.

Minor comment on tests, otherwise good to go in.

**Comment by IgnaceBleukx** [comment]
This actually also closes one of our oldest issues: #124

**Comment by IgnaceBleukx** [comment]
There is an edge case apparently where a np array can have 0 dimensions: e.g., `np.array(3)` is a valid Numpy array, which causes our `_vectorized` helper function to fail (calls `len` on the array).
Simple fix would be to add 

```python
if isinstance(other, np.ndarray) and other.ndim == 0:
    other = other.item() if other.dtype != object else other[()]
if self.ndim == 0:
    s = self.item() if self.dtype == object else self[()]
    return getattr(s, attr)(other)
```

but it introduced another isinstance check... How shall we continue?

**Comment by tias** [comment]
all this shows is that we are increasingly creeping up to re-implementing what ufunc would all handle...

also in the typing, there are more edge cases: we should also allow plain python lists for consistency, and probably smth else, and smth else...

So I've removed the type annotations again (they are incomplete, as is the code handling).


I'm OK to put this PR in, I'm also OK for you to cover the scalar case. But we have to draw a line for this release somewhere closeby.

And the next phase is just proper ufunc.

**Comment by IgnaceBleukx** [comment]
I'm ok leaving the edge case for now, I cannot think of an actual use case where you would end up with such an "array", definitely not in common modeling. So indeed let's just draw the line here for this release, and v0.21 can have proper ufuncs
